### PR TITLE
overlord/ifacestate: no interface checks if no snap id

### PR DESCRIPTION
**Pull request by chipaca**

### CODE REVIEW

File: overlord/ifacestate/helpers.go [hunk 1]
```diff
@@ -297,18 +297,19 @@ func CheckInterfaces(st *state.State, snapInfo *snap.Info) error {
 	// XXX: AddImplicitSlots is really a brittle interface
 	snap.AddImplicitSlots(snapInfo)
 
+	if snapInfo.SnapID == "" {
+		// no SnapID means --dangerous was given, so skip interface checks
```
Comment (mvo5) [inline overlord/ifacestate/helpers.go]: Hm, `sudo snap install --dangerous hello` has a SnapID in my state.json file. So the comment is slightly misleading (at least to me). Are we actually testing for local (sideload) installs? If so, would it make sense to check for snapInfo.Revision.Local() here?

Comment (pedronis) [inline overlord/ifacestate/helpers.go]: @mvo5 dangerous doesn't mean anything when using the store and not a local snap, maybe we should error if it's given in that case (only the local snap case in api.go looks at dangerous)

Comment (niemeyer) [inline overlord/ifacestate/helpers.go]: Yeah, it's probably a good idea to error. We're not really doing anything useful when it comes from the store at the moment.

File: overlord/ifacestate/helpers.go [hunk 2]
```diff
@@ -191,24 +191,21 @@ func (c *autoConnectChecker) snapDeclaration(snapID string) (*asserts.SnapDeclar
...
 func (c *autoConnectChecker) check(plug *interfaces.Plug, slot *interfaces.Slot) bool {
-	var plugDecl *asserts.SnapDeclaration
-	if plug.Snap.SnapID != "" {
-		var err error
-		plugDecl, err = c.snapDeclaration(plug.Snap.SnapID)
-		if err != nil {
-			logger.Noticef("error: cannot find snap declaration for %q: %v", plug.Snap.Name(), err)
-			return false
-		}
+	if plug.Snap.SnapID == "" || slot.Snap.SnapID == "" {
```
Comment (pedronis) [inline overlord/ifacestate/helpers.go]: we can't afford to do this unless going forward,  we separate sensible-to-auto-connect vs allow-to-auto-connect

File: overlord/ifacestate/handlers.go [hunk 3]
```diff
@@ -334,9 +334,11 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
 		BaseDeclaration:     baseDecl,
 	}
 
-	err = ic.Check()
-	if err != nil {
-		return err
+	if plugDecl != nil && slotDecl != nil {
```
Comment (pedronis) [inline overlord/ifacestate/handlers.go]: a comment before this would be in order I think

File: overlord/ifacestate/ifacemgr_test.go [hunk 4]
```diff
@@ -706,6 +708,15 @@ func (s *interfaceManagerSuite) TestDoSetupSnapSecurityAutoConnects(c *C) {
 
 // The setup-profiles task will auto-connect plugs with viable candidates also condidering snap declarations.
 func (s *interfaceManagerSuite) TestDoSetupSnapSecurityAutoConnectsDeclBased(c *C) {
+	s.testDoSetupSnapSecurityAutoConnectsDeclBased(c, true)
+}
+
+// The setup-profiles task will auto-connect plugs with viable candidates when snap declarations are missing.
+func (s *interfaceManagerSuite) TestDoSetupSnapSecurityAutoConnectsDeclBasedWhenMissingDecl(c *C) {
```
Comment (pedronis) [inline overlord/ifacestate/ifacemgr_test.go]: isn't this what we aren't doing now?

Comment (pedronis) [inline overlord/ifacestate/ifacemgr_test.go]: or it the comment that is wrong now?

File: overlord/ifacestate/ifacemgr_test.go [hunk 5]
```diff
@@ -747,19 +760,26 @@ slots:
...
+		// Ensure that "test" plug is now saved in the state as auto-connected.
+		c.Check(conns, DeepEquals, map[string]interface{}{
+			"consumer:plug producer:slot": map[string]interface{}{"auto": true, "interface": "test"},
+		})
+		// Ensure that "test" is really connected.
+		c.Check(plug.Connections, HasLen, 1)
+	} else {
+		// Ensure nothing is connected.
+		c.Check(conns, HasLen, 0)
+		c.Check(plug.Connections, HasLen, 0)
```
Comment (pedronis) [inline overlord/ifacestate/ifacemgr_test.go]: having the checks moved here is a bit obscure I think

File: overlord/ifacestate/ifacemgr_test.go (hunk 4 above)
Comment (chipaca) [inline overlord/ifacestate/ifacemgr_test.go]: the comment is wrong now :-(

**Pull request merged**